Recast/Source/RecastMeshDetail.cpp

```diff
@@ -211,24 +211,56 @@ static unsigned short getHeight(const float fx, const float fy, const float fz,
 	unsigned short h = hp.data[ix+iz*hp.width];
 	if (h == RC_UNSET_HEIGHT)
 	{
+		// Search in a spiral
+		const int rad = rcMax(1, (int)ceilf(maxEdgeError));
+
+		int x = 1, z = 0, dx = 1, dz = 0;
+		int maxSize = rad * 2 + 1;
+		int maxIter = maxSize * maxSize - 1;
+		int nextRingIterStart = 8;
+		int nextRingIters = 16;
+
 		// Special case when data might be bad.
 		// Find nearest neighbour pixel which has valid height.
-		const int off[8*2] = { -1,0, -1,-1, 0,-1, 1,-1, 1,0, 1,1, 0,1, -1,1};
 		float dmin = FLT_MAX;
-		for (int i = 0; i < 8; ++i)
+		for (int i = 0; i < maxIter; i++)
 		{
-			const int nx = ix+off[i*2+0];
-			const int nz = iz+off[i*2+1];
-			if (nx < 0 || nz < 0 || nx >= hp.width || nz >= hp.height) continue;
-			const unsigned short nh = hp.data[nx+nz*hp.width];
-			if (nh == RC_UNSET_HEIGHT) continue;
-			
-			const float d = fabsf(nh*ch - fy);
-			if (d < dmin)
+			const int nx = ix + x;
+			const int nz = iz + z;
+
+			if (nx >= 0 && nz >= 0 && nx < hp.width && nz < hp.height)
+			{
+				const unsigned short nh = hp.data[nx + nz*hp.width];
+				if (nh != RC_UNSET_HEIGHT)
+				{
+					const float d = fabsf(nh*ch - fy);
+					if (d < dmin)
+					{
+						h = nh;
+						dmin = d;
+					}
+				}
+			}
+
+			// Abort the search if we're entering a new 'ring' and have found a result
+			if (i + 1 == nextRingIterStart)
+			{
+				if (h != RC_UNSET_HEIGHT)
+					break;
+
+				nextRingIterStart += nextRingIters;
+				// Each successive 'ring' has 8 additional cells in it
```

Comment:
The ring and 8 does not make sense, can you elaborate?
I think the reference to the spiral loop is from here (I thought it was usenet thing, but nope):
http://stackoverflow.com/questions/398299/looping-in-a-spiral